### PR TITLE
chore: Introduce fiber stack allocator

### PR DESCRIPTION
**Pull request by romange**
1. Use mi_malloc for allocating fiber stacks. Introduce FiberStackResource - a memory resource dedicated for handling stack allocations, and inject it into Fibers library as well as accept listeners.

2. Add "memory_fiberstack_vms_bytes" metric exposing fiber stack vm usage.

### CODE REVIEW

File: src/core/mi_memory_resource.cc [hunk 1]
```diff
@@ -3,18 +3,25 @@
...
+
 #include "base/logging.h"
 
 namespace dfly {
 
-void* MiMemoryResource::do_allocate(std::size_t size, std::size_t align) {
+using namespace std;
+
+inline bool is_page_aligned(void* ptr) {
+  return (uintptr_t(ptr) & 4095) == 0;
```
Comment (BorysTheDev) [inline src/core/mi_memory_resource.cc:15]: maybe add a parameter for huge pages too?

File: src/core/mi_memory_resource.cc [hunk 2]
```diff
@@ -40,4 +47,37 @@ void MiMemoryResource::do_deallocate(void* ptr, std::size_t size, std::size_t al
...
+  // deliberately ignore alignment because mimalloc overprovisions memory by align -1 bytes.
+  void* res = mi_malloc(size);
+  if (res == nullptr) {
+    throw bad_alloc();
+  }
+
+  // in theory res does not have to be page aligned. In that case madvise will silently fail.
+  DCHECK(is_page_aligned(res)) << res;
+
+  madvise((void*)res, 4096, MADV_DONTNEED);
```
Comment (BorysTheDev) [inline src/core/mi_memory_resource.cc:60]: I suggest to provide ability to choose between normal and huge pages

File: src/server/dfly_main.cc [hunk 3]
```diff
@@ -690,10 +699,14 @@ Usage: dragonfly [FLAGS]
     LOG(INFO) << "Max memory limit is: " << hr_limit;
   }
 
+  // Initialize mi_malloc options
+  // export MIMALLOC_VERBOSE=1 to see the options before the override.
   mi_option_enable(mi_option_show_errors);
   mi_option_set(mi_option_max_warnings, 0);
   mi_option_set(mi_option_decommit_delay, 1);
 
+  fb2::SetDefaultStackResource(&FiberStackResource::instance(), 32 * 1024);
```
Comment (BorysTheDev) [inline src/server/dfly_main.cc:708]: I'd prefer to create a constant

File: src/core/mi_memory_resource.cc [hunk 4]
```diff
@@ -40,4 +47,37 @@ void MiMemoryResource::do_deallocate(void* ptr, std::size_t size, std::size_t al
...
+
+  DVLOG(1) << "Allocate [" << res << " " << (char*)res + size << ") size=" << size
+           << " align=" << align << " usable=" << mi_usable_size(res);
+  return res;
+}
+
+void FiberStackResource::do_deallocate(void* ptr, size_t size, size_t align) {
+  DCHECK_GT(size, 8192u);
```
Comment (BorysTheDev) [inline src/core/mi_memory_resource.cc:68]: could you add comment why

**Review by BorysTheDev — APPROVED** [review]

File: src/core/mi_memory_resource.cc (hunk 1 above)
Comment (romange) [reply src/core/mi_memory_resource.cc:15]: VM pages are not part of this abstraction layer, it's something that OS manages underneath but I think you are right, if the page backing the stack is a large page, I think madvise will fail. i will check it

File: src/core/mi_memory_resource.cc (hunk 2 above)
Comment (romange) [reply src/core/mi_memory_resource.cc:60]: see my previous point, we do not control this, it's hidden by mi_malloc.

**Comment by romange** [comment]
@BorysTheDev  I removed madvise because it does not work with large pages, obviously we can not unload parts of a  page. 
I am putting the PR on hold till next week.

**Review by BorysTheDev — APPROVED** [review]

**Pull request merged**